File: Examples/ReferenceDataRequest.cs
//------------------------------------------------------------------------------
// <copyright project="Examples" file="ReferenceDataRequest.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Examples
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    using BEmu; //un-comment this line to use the Bloomberg API Emulator
    //using Bloomberglp.Blpapi; //un-comment this line to use the actual Bloomberg API

    public static class ReferenceDataRequest
    {
        /// <summary>
        /// This example Bloomberg request starts a session meant for Reference Requests and requests a few fields for several securities.
        /// I pulled this example code almost line-for-line from section C.1 of the Bloomberg API Developer's Guide
        /// </summary>
        public static void RunExample()
        {
            SessionOptions sessionOptions = new SessionOptions();
            sessionOptions.ServerHost = "localhost";
            sessionOptions.ServerPort = 8194;

            Session session = new Session(sessionOptions);
            if (session.Start() && session.OpenService("//blp/refdata"))
            {
                Service refDataSvc = session.GetService("//blp/refdata");
                if (refDataSvc == null)
                {
                    Console.WriteLine("Cannot get service");
                }
                else
                {
                    CorrelationID requestID = new CorrelationID(1);
                    Request request = refDataSvc.CreateRequest("ReferenceDataRequest");

                    //request information for the following securities
                    request.Append("securities", "SPY US EQUITY");
                    request.Append("securities", "MSFT US EQUITY");
                    request.Append("securities", "AAPL 150117C00600000 EQUITY"); //this is a stock option: TICKER yyMMdd[C/P]\d{8} EQUITY
                    //request.Append("securities", "ZYZZ US EQUITY"); //the code treats securities that start with a "Z" as non-existent

                    //include the following simple fields in the result
                    //request.Append("fields", "ZPX_LAST"); //the code treats a field that starts with a "Z" as a bad field
                    request.Append("fields", "PX_LAST");
                    request.Append("fields", "BID");
                    request.Append("fields", "ASK");
                    request.Append("fields", "TICKER");
                    request.Append("fields", "OPT_EXPIRE_DT"); //only stock options have this field

                    //request a field that can be overriden and returns bulk data
                    request.Append("fields", "CHAIN_TICKERS"); //only stocks have this field 
                    Element overrides = request["overrides"];

                    //request only puts
                    Element ovrdPutCall = overrides.AppendElement();
                    ovrdPutCall.SetElement("fieldId", "CHAIN_PUT_CALL_TYPE_OVRD");
                    ovrdPutCall.SetElement("value", "P"); //accepts either "C" for calls or "P" for puts

                    //request 5 options in the result
                    Element ovrdNumStrikes = overrides.AppendElement();
                    ovrdNumStrikes.SetElement("fieldId", "CHAIN_POINTS_OVRD");
                    ovrdNumStrikes.SetElement("value", 5); //accepts a positive integer

                    //request options that expire on Dec. 20, 2014
                    Element ovrdDtExps = overrides.AppendElement();
                    ovrdDtExps.SetElement("fieldId", "CHAIN_EXP_DT_OVRD");
                    ovrdDtExps.SetElement("value", "20141220"); //accepts dates in the format yyyyMMdd (this is Dec. 20, 2014)

                    session.SendRequest(request, requestID);

                    bool continueToLoop = true;
                    while (continueToLoop)
                    {
                        Event eventObj = session.NextEvent();
                        switch (eventObj.Type)
                        {
                            case Event.EventType.RESPONSE: // final event
                                continueToLoop = false;
                                ReferenceDataRequest.handleResponseEvent(eventObj);
                                break;
                            case Event.EventType.PARTIAL_RESPONSE:
                                ReferenceDataRequest.handleResponseEvent(eventObj);
                                break;
                            default:
                                ReferenceDataRequest.handleOtherEvent(eventObj);
                                break;
                        }
                    }
                }
            }
            else
            {
                Console.WriteLine("Cannot connect to server.  Check that the server host is \"localhost\" or \"127.0.0.1\" and that the server port is 8194.");
            }
        }

        private static void handleResponseEvent(Event eventObj)
        {
            Console.WriteLine();
            Console.WriteLine();
            Console.WriteLine("EventType =" + eventObj.Type);
            foreach (Message message in eventObj.GetMessages())
            {
                Console.WriteLine();
                Console.WriteLine();
                Console.WriteLine("correlationID=" + message.CorrelationID);
                Console.WriteLine("messageType =" + message.MessageType);

                Element elmSecurityDataArray = message["securityData"];

                for (int valueIndex = 0; valueIndex < elmSecurityDataArray.NumValues; valueIndex++)
                {
                    Element elmSecurityData = elmSecurityDataArray.GetValueAsElement(valueIndex);

                    string security = elmSecurityData.GetElementAsString("security");
                    Console.WriteLine(security);

                    bool hasFieldErrors = elmSecurityData.HasElement("fieldExceptions", true);
                    if (hasFieldErrors)
                    {
                        Element elmFieldErrors = elmSecurityData["fieldExceptions"];
                        for (int errorIndex = 0; errorIndex < elmFieldErrors.NumValues; errorIndex++)
                        {
                            Element fieldError = elmFieldErrors.GetValueAsElement(errorIndex);
                            string fieldId = fieldError.GetElementAsString("fieldId");

                            Element errorInfo = fieldError["errorInfo"];
                            string source = errorInfo.GetElementAsString("source");
                            int code = errorInfo.GetElementAsInt32("code");
                            string category = errorInfo.GetElementAsString("category");
                            string strMessage = errorInfo.GetElementAsString("message");
                            string subCategory = errorInfo.GetElementAsString("subcategory");

                            Console.Error.WriteLine();
                            Console.Error.WriteLine();
                            Console.Error.WriteLine("\tfield error: " + security);
                            Console.Error.WriteLine(string.Format("\tfieldId = {0}", fieldId));
                            Console.Error.WriteLine(string.Format("\tsource = {0}", source));
                            Console.Error.WriteLine(string.Format("\tcode = {0}", code));
                            Console.Error.WriteLine(string.Format("\tcategory = {0}", category));
                            Console.Error.WriteLine(string.Format("\terrorMessage = {0}", strMessage));
                            Console.Error.WriteLine(string.Format("\tsubCategory = {0}", subCategory));
                        }
                    }

                    bool isSecurityError = elmSecurityData.HasElement("securityError", true);
                    if (isSecurityError)
                    {
                        Element secError = elmSecurityData["securityError"];
                        string source = secError.GetElementAsString("source");
                        int code = secError.GetElementAsInt32("code");
                        string category = secError.GetElementAsString("category");
                        string errorMessage = secError.GetElementAsString("message");
                        string subCategory = secError.GetElementAsString("subcategory");

                        Console.Error.WriteLine();
                        Console.Error.WriteLine();
                        Console.Error.WriteLine("\tsecurity error");
                        Console.Error.WriteLine(string.Format("\tsource = {0}", source));
                        Console.Error.WriteLine(string.Format("\tcode = {0}", code));
                        Console.Error.WriteLine(string.Format("\tcategory = {0}", category));
                        Console.Error.WriteLine(string.Format("\terrorMessage = {0}", errorMessage));
                        Console.Error.WriteLine(string.Format("\tsubCategory = {0}", subCategory));
                    }
                    else
                    {
                        Element elmFieldData = elmSecurityData["fieldData"];

                        double pxLast = elmFieldData.GetElementAsFloat64("PX_LAST");
                        double bid = elmFieldData.GetElementAsFloat64("BID");
                        double ask = elmFieldData.GetElementAsFloat64("ASK");
                        string ticker = elmFieldData.GetElementAsString("TICKER");

                        Console.WriteLine();
                        Console.WriteLine();
                        Console.WriteLine("\tPX_LAST = " + pxLast.ToString());
                        Console.WriteLine("\tBID = " + bid.ToString());
                        Console.WriteLine("\tASK = " + ask.ToString());
                        Console.WriteLine("\tTICKER = " + ticker.ToString());

                        bool excludeNullElements = true;
                        if (elmFieldData.HasElement("CHAIN_TICKERS", excludeNullElements)) //be careful, the excludeNullElements argument is false by default
                        {
                            Element chainTickers = elmFieldData["CHAIN_TICKERS"];
                            for (int chainTickerValueIndex = 0; chainTickerValueIndex < chainTickers.NumValues; chainTickerValueIndex++)
                            {
                                Element chainTicker = chainTickers.GetValueAsElement(chainTickerValueIndex);
                                string strChainTicker = chainTicker.GetElementAsString("Ticker");

                                Console.WriteLine("\tCHAIN_TICKER = " + strChainTicker.ToString());
                            }
                        }
                        else
                        {
                            Console.WriteLine("\tNo CHAIN_TICKER information");
                        }
                    }

                }
            }
        }

        private static void handleOtherEvent(Event eventObj)
        {
            Console.WriteLine("EventType=" + eventObj.Type);
            foreach (Message message in eventObj.GetMessages())
            {
                Console.WriteLine("correlationID=" + message.CorrelationID);
                Console.WriteLine("messageType=" + message.MessageType);
                Console.WriteLine(message.ToString());
                if (Event.EventType.SESSION_STATUS == eventObj.Type && message.MessageType.Equals("SessionTerminated"))
                {
                    Console.WriteLine("Terminating: " + message.MessageType);
                }
            }
        }
    }
}

File: Examples_cpp/src/ReferenceDataRequest.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/ReferenceDataRequest.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include <iostream>
#include <time.h>
#include "api_selector.h"

using namespace std;

namespace Examples
{
	void ProcessReferenceDataEvent(Event evt);

	void RunReferenceDataExample()
	{
		SessionOptions soptions;
		soptions.setServerHost("localhost");
		soptions.setServerPort(8194);

		Session session(soptions);
		if(session.start() && session.openService("//blp/refdata"))
		{
			Service refDataSvc = session.getService("//blp/refdata");

            CorrelationId requestID(1); //1 is just some number I picked
            Request request = refDataSvc.createRequest("ReferenceDataRequest");

            //request information for the following securities
            request.append("securities", "SPY US EQUITY");
            request.append("securities", "MSFT US EQUITY");
            request.append("securities", "AAPL 150117C00600000 EQUITY"); //this is a stock option: TICKER yyMMdd[C/P]\d{8} EQUITY
            //request.append("securities", "ZYZZ US EQUITY"); //the code treats securities that start with a "Z" as non-existent

            //include the following simple fields in the result
            //request.append("fields", "ZPX_LAST"); //the code treats a field that starts with a "Z" as a bad field
            request.append("fields", "PX_LAST");
            request.append("fields", "BID");
            request.append("fields", "ASK");
            request.append("fields", "TICKER");
            request.append("fields", "OPT_EXPIRE_DT"); //only stock options have this field

            //request a field that can be overriden and returns bulk data
            request.append("fields", "CHAIN_TICKERS"); //only stocks have this field 
			Element overrides = request.getElement("overrides");
			
            //request only puts
			Element ovrdPutCall = overrides.appendElement();
            ovrdPutCall.setElement("fieldId", "CHAIN_PUT_CALL_TYPE_OVRD");
            ovrdPutCall.setElement("value", "P"); //accepts either "C" for calls or "P" for puts

            //request 5 options in the result
            Element ovrdNumStrikes = overrides.appendElement();
            ovrdNumStrikes.setElement("fieldId", "CHAIN_POINTS_OVRD");
            ovrdNumStrikes.setElement("value", 5); //accepts a positive integer

            //request options that expire on Dec. 20, 2014
            Element ovrdDtExps = overrides.appendElement();
            ovrdDtExps.setElement("fieldId", "CHAIN_EXP_DT_OVRD");
            ovrdDtExps.setElement("value", "20141220"); //accepts dates in the format yyyyMMdd (this is Dec. 20, 2014)

            session.sendRequest(request, requestID);

            bool continueToLoop = true;
            while (continueToLoop)
            {
                Event evt = session.nextEvent();
				switch (evt.eventType())
                {
					case Event::RESPONSE:
						continueToLoop = false; //fall through
					case Event::PARTIAL_RESPONSE:
						ProcessReferenceDataEvent(evt);
						break;
                }
            }
		}
		else
		{
			cerr << "Cannot connect to server";
		}
	}

	void ProcessReferenceDataEvent(Event evt)
	{
		const string level1 = "";
		const string level2 = "\t";
		const string level3 = "\t\t";
		const string level4 = "\t\t\t";

		cout << endl << endl;
		cout << level1 << "EventType = " << evt.eventType();

		MessageIterator iter(evt);
		while(iter.next())
		{
			Message msg = iter.message();

			cout << endl << endl;
			cout << level1 << "correlationID = " << msg.correlationId().asInteger() << endl;
			cout << level1 << "messageType = " << msg.messageType().string() << endl;

			cout << endl << endl;
			Element elmSecurityDataArray = msg.getElement("securityData");
			
			for (size_t valueIndex = 0; valueIndex < elmSecurityDataArray.numValues(); valueIndex++)
			{
				Element elmSecurityData = elmSecurityDataArray.getValueAsElement(valueIndex);

				string security = elmSecurityData.getElementAsString("security");
				cout << level2 << security << endl;

				bool hasFieldErrors = elmSecurityData.hasElement("fieldExceptions", true);
				if (hasFieldErrors)
				{
					Element elmFieldErrors = elmSecurityData.getElement("fieldExceptions");
					for (size_t errorIndex = 0; errorIndex < elmFieldErrors.numValues(); errorIndex++)
					{
						Element fieldError = elmFieldErrors.getValueAsElement(errorIndex);
						string fieldId = fieldError.getElementAsString("fieldId");

						Element errorInfo = fieldError.getElement("errorInfo");
                        string source = errorInfo.getElementAsString("source");
                        int code = errorInfo.getElementAsInt32("code");
                        string category = errorInfo.getElementAsString("category");
                        string strMessage = errorInfo.getElementAsString("message");
                        string subCategory = errorInfo.getElementAsString("subcategory");

						cerr << level3 << "field error:" << endl;
						cerr << level4 << "fieldId = " << fieldId << endl;
						cerr << level4 << "source = " << source << endl;
						cerr << level4 << "code = " << code << endl;
						cerr << level4 << "category = " << category << endl;
						cerr << level4 << "errorMessage = " << strMessage << endl;
						cerr << level4 << "subCategory = " << subCategory << endl;
					}
				}

				bool isSecurityError = elmSecurityData.hasElement("securityError", true);
				if (isSecurityError)
				{
					Element secError = elmSecurityData.getElement("securityError");
                    string source = secError.getElementAsString("source");
                    int code = secError.getElementAsInt32("code");
                    string category = secError.getElementAsString("category");
                    string errorMessage = secError.getElementAsString("message");
                    string subCategory = secError.getElementAsString("subcategory");

					cerr << level3 << "security error:" << endl;
					cerr << level4 << "source = " << source << endl;
					cerr << level4 << "code = " << code << endl;
					cerr << level4 << "category = " << category << endl;
					cerr << level4 << "errorMessage = " << errorMessage << endl;
					cerr << level4 << "subCategory = " << subCategory << endl;
				}
				else
				{
					Element elmFieldData = elmSecurityData.getElement("fieldData");

                    double pxLast = elmFieldData.getElementAsFloat64("PX_LAST");
                    double bid = elmFieldData.getElementAsFloat64("BID");
                    double ask = elmFieldData.getElementAsFloat64("ASK");
                    string ticker = elmFieldData.getElementAsString("TICKER");
					
					cout << level3 << "fields: " << endl;
                    cout << level4 << "PX_LAST = " << pxLast << endl;
                    cout << level4 << "BID = " << bid << endl;
                    cout << level4 << "ASK = " << ask << endl;
                    cout << level4 << "TICKER = " << ticker << endl;

					bool excludeNullElements = true;
					if (elmFieldData.hasElement("CHAIN_TICKERS", excludeNullElements)) //be careful, the excludeNullElements argument is false by default
					{
						Element chainTickers = elmFieldData.getElement("CHAIN_TICKERS");
						for (size_t chainTickerValueIndex = 0; chainTickerValueIndex < chainTickers.numValues(); chainTickerValueIndex++)
						{
                            Element chainTicker = chainTickers.getValueAsElement(chainTickerValueIndex);
                            string strChainTicker = chainTicker.getElementAsString("Ticker");

							cout << level4 << "CHAIN_TICKER = " << strChainTicker << endl;
						}
					}
					else
					{
						cout << level4 << "No CHAIN_TICKER information" << endl;
					}


				}

			}

		}

	}
}